core/trino-main/src/main/java/io/trino/execution/buffer/PagesSerdeUtil.java

```diff
@@ -124,29 +132,44 @@ protected Page computeNext()
         }
     }
 
-    public static Iterator<Slice> readSerializedPages(SliceInput sliceInput)
+    public static Iterator<Slice> readSerializedPages(InputStream inputStream)
     {
-        return new SerializedPageReader(sliceInput);
+        return new SerializedPageReader(inputStream);
     }
 
     private static class SerializedPageReader
             extends AbstractIterator<Slice>
     {
-        private final SliceInput input;
+        private final InputStream inputStream;
+        private final byte[] headerBuffer = new byte[SERIALIZED_PAGE_HEADER_SIZE];
+        private final Slice headerSlice = Slices.wrappedBuffer(headerBuffer);
 
-        SerializedPageReader(SliceInput input)
+        SerializedPageReader(InputStream input)
         {
-            this.input = requireNonNull(input, "input is null");
+            this.inputStream = requireNonNull(input, "inputStream is null");
         }
 
         @Override
         protected Slice computeNext()
         {
-            if (!input.isReadable()) {
-                return endOfData();
+            try {
+                int read = inputStream.read(headerBuffer);
```

Comment:
Technically the read method is allowed to return 0 or any number of bytes actually read lower than the requested length, even without reaching the end of stream. It’s a specific behavior of ByteStreams.read to stop either at EOS or fully read the requested length by looping until either condition is satisfied.

Reply:
Oh you are right, I didn't realize the difference here. Thanks for catching it.